src/feature/hs/hs_descriptor.c

```diff
@@ -1182,6 +1182,38 @@ desc_encode_v3(const hs_descriptor_t *desc,
 
 /* === DECODING === */
 
+/* Given the token tok for an auth client, decode it as
+ * hs_desc_authorized_client_t. tok->args MUST contain at least 3 elements
+ * Return 0 on success else -1 on failure. */
+static int
+decode_auth_client(const directory_token_t *tok,
+                   hs_desc_authorized_client_t *client)
+{
+  int ret = -1;
+  tor_assert(tok);
+  tor_assert(tok->n_args >= 3);
+  tor_assert(client);
+
+  if (base64_decode((char *) client->client_id, sizeof(client->client_id),
```

Comment:
It's not enough to check that this didn't return a negative number -- you need to make sure that it returned the right positive number, or that its input was the right length.

**Please audit all of the base*_decode() functions in this branch to make sure that they don't have this bug; I may have missed some.**

Reply:
Fixup commit: 97404e3a5c